[MRG] MNT: Drop boilerplate fused type `typedef`s

jakirkham opened the pull request:
When generating C code from Cython without this code, Cython seems able to figure out that function pointers are needed and what kinds within each function where we assign them. So going through the effort of forward declaring fused type versions of them and declaring the function pointers in each function is unneeded. Not to mention that some of the fused types that were used later are not declared with `cdef`s nor are `ctypedef`s for them above using fused types. Given all of this, it makes more sense to remove this unneeded boilerplate instead of updating it to include the missing definitions and declarations.

<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->


#### What does this implement/fix? Explain your changes.


#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

rth reviewed (approved):
I'm not exactly sure what Cython is doing in this case -- the generated C code doesn't appear to change that much in any case (cf [cd_fast_master.html.gz](https://github.com/scikit-learn/scikit-learn/files/2193121/cd_fast_master.html.gz) and [cd_fast_PR.html.gz](https://github.com/scikit-learn/scikit-learn/files/2193122/cd_fast_PR.html.gz))

LGTM assuming tests pass, thanks @jakirkham !

jakirkham commented:
It's not very obvious from the HTML file, but if one runs `cython sklearn/linear_model/cd_fast.pyx`, it will show something like this in each specialization of the function. Dropping the declaration in Cython has no real effect on the pattern seen below as Cython figures out the type definition and adds it in either case.

```C
static PyObject *__pyx_pf_7sklearn_12linear_model_7cd_fast_8enet_coordinate_descent(CYTHON_UNUSED PyObject *__pyx_self, PyArrayObject *__pyx_v_w, float __pyx_v_alpha, float __pyx_v_beta, PyArrayObject *__pyx_v_X, PyArrayObject *__pyx_v_y, int __pyx_v_max_iter, float __pyx_v_tol, PyObject *__pyx_v_rng, int __pyx_v_random, int __pyx_v_positive) {
    ...
    float (*__pyx_v_dot)(int, float *, int, float *, int);
    ...
    __pyx_v_dot = cblas_sdot;
    ...
    __pyx_v_tol = (__pyx_v_tol * __pyx_v_dot(__pyx_v_n_samples, __pyx_v_y_data, __pyx_v_n_tasks, __pyx_v_y_data, __pyx_v_n_tasks));
    ...
}
```

IOW both cases result in the definition of a function pointer in the scope of the C function. The function pointer is then assigned a particular function based on the fused type specialization. That function pointer is called in various points afterwards.

AFAICT the only value in declaring this function pointer earlier is either to aid the reader (though it is pretty clear anyways) or aid Cython (though it figures this out just fine).

jakirkham committed:
MNT: Drop boilerplate fused type `typedef`s

When generating C code from Cython without this code, Cython seems able
to figure out that function pointers are needed and what kinds within
each function where we assign them. So going through the effort of
forward declaring fused type versions of them and declaring the function
pointers in each function is unneeded. Not to mention that some of the
fused types that were used later are not declared with `cdef`s nor are
`ctypedef`s for them above using fused types. Given all of this, it
makes more sense to remove this unneeded boilerplate instead of updating
it to include the missing definitions and declarations.

jakirkham commented:
> The same pattern is also used in `sklearn/cluster/_k_means.pyx`: would you mind also applying these changes there?

Sure, done. Grepped through the code for other cases, but didn't see any.

rth commented:
Thanks for the clarifications and the additional fixes.

(yes, I should have just run a diff on the generated c files instead of looking at the HTML output -- that would have been simpler).
